lib/path-design.md

```diff
@@ -0,0 +1,177 @@
+
+<details>
+<summary>Arguments</summary>
+
+- :heavy_plus_sign: It enables the law that if `normalise p == normalise q` then `$(stat p) == $(stat q)`.
+- Comparison to other frameworks to figure out the least surprising behavior:
+  - :heavy_plus_sign: Nix itself doesn't preserve trailing newlines when parsing and appending its paths
+  - :heavy_minus_sign: [Rust's std::path](https://doc.rust-lang.org/std/path/index.html) does preserve them during [construction](https://doc.rust-lang.org/std/path/struct.Path.html#method.new)
+    - :heavy_plus_sign: Doesn't preserve them when returning individual [components](https://doc.rust-lang.org/std/path/struct.Path.html#method.components)
+    - :heavy_plus_sign: Doesn't preserve them when [canonicalizing](https://doc.rust-lang.org/std/path/struct.Path.html#method.canonicalize)
+  - :heavy_plus_sign: [Python 3's pathlib](https://docs.python.org/3/library/pathlib.html#module-pathlib) doesn't preserve them during [construction](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath)
+    - Notably it represents the individual components as a list internally
+  - :heavy_minus_sign: [Haskell's filepath](https://hackage.haskell.org/package/filepath-1.4.100.0) has [explicit support](https://hackage.haskell.org/package/filepath-1.4.100.0/docs/System-FilePath.html#g:6) for handling trailing slashes
+    - :heavy_minus_sign: Does preserve them for [normalisation](https://hackage.haskell.org/package/filepath-1.4.100.0/docs/System-FilePath.html#v:normalise)
+  - :heavy_minus_sign: [NodeJS's Path library](https://nodejs.org/api/path.html) preserves trailing slashes for [normalisation](https://nodejs.org/api/path.html#pathnormalizepath)
+    - :heavy_plus_sign: For [parsing a path](https://nodejs.org/api/path.html#pathparsepath) into its significant elements, trailing slashes are not preserved
+- :heavy_plus_sign: Nix's builtin function `dirOf` gives an unexpected result for paths with trailing slashes: `dirOf "foo/bar/" == "foo/bar"`.
+  Inconsistently, `baseNameOf` works correctly though: `baseNameOf "foo/bar/" == "bar"`.
+  - :heavy_minus_sign: We are writing a path library to improve handling of paths though, so we shouldn't use these functions and discourage their use
+- :heavy_minus_sign: Unexpected result when normalising intermediate paths, like `normalise ("foo" + "/") + "bar" == "foobar"`
+  - :heavy_plus_sign: Does this have a real use case?
+  - :heavy_plus_sign: Don't use `+` to append paths, this library has a `join` function for that
+    - :heavy_minus_sign: Users might use `+` out of habit though
+- :heavy_plus_sign: The `realpath` command also removes trailing slashes
+- :heavy_plus_sign: Even with a trailing slash, the path is the same, it's only an indication that it's a directory
+- :heavy_plus_sign: Normalisation should return the same string when we know it's the same path, so removing the slash.
+  This way we can use the result as an attribute key.
```

Comment:
I removed this in a commit now, but the idea was that e.g.

```nix
{
  environment.etc."some/file.conf".source = foo;
}
```

and

```nix
{
  environment.etc."some/file.conf/".source = bar;
}
```

should trigger an eval-time error because the same path is assigned twice.